### PR TITLE
Update PropertyM Applicative/Monad instance

### PR DESCRIPTION
**Pull request by gliptak**
Signed-off-by: Gábor Lipták <gliptak@gmail.com>

https://github.com/nick8325/quickcheck/issues/127

### CODE REVIEW

File: Test/QuickCheck/Monadic.hs [hunk 1]
```diff
@@ -105,14 +108,23 @@ newtype PropertyM m a =
...
-instance Monad m => Applicative (PropertyM m) where
-  pure = return
-  (<*>) = liftM2 ($)
+instance Applicative (PropertyM m) where
+  pure x  = MkPropertyM ($ x)
+  f <*> v = MkPropertyM $ \ c -> unPropertyM f $ \ g -> unPropertyM v (c . g)
+
+#if MIN_VERSION_base(4,9,0)
+instance Monad (PropertyM m) where
+  return x = MkPropertyM ($ x)
```
Comment (phadej) [inline Test/QuickCheck/Monadic.hs]: just `return = pure` would be clearer? Or actually omitting it fully in base-4.9 case?

Comment (gliptak) [inline Test/QuickCheck/Monadic.hs]: @Icelandjack Would you comment? Thanks

**Comment by nick8325** [comment]
Thanks for the patch! There is a bit of duplicated code here, which I would like to avoid. In particular:

* `<*>` and `>>=` share a lot of the same code. To avoid this, we could define `bind :: PropertyM m a -> (a -> PropertyM m b) -> PropertyM m b` and use that to define `<*>` and `>>=`.
* The `Monad` instance for `PropertyM` appears twice (with slightly different code). I would like to keep the existing definitions as they work fine. Instead of duplicating it we can use CPP to a) remove the `Monad m` superclass constraint, and b) add the line `instance Monad m => MonadFail (PropertyM m) where` before the definition of `fail`.
* The duplication between `return` and `pure` that @phadej pointed out.

Also, please could you not use `MIN_VERSION_base` but make the cabal file define a preprocessor macro, e.g. `NO_MONAD_FAIL`, on base < 4.9. This makes it possible to load QuickCheck directly into GHCi which is handy for development/debugging.

**Comment by gliptak** [comment]
@nick8325 Thank you for helping out the refactoring

**Pull request closed without merging**